keyword.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	jptr "github.com/qri-io/jsonpointer"
+)
+
+var notSupported = map[string]bool{
+	// "$schema": true,
+	// "$id": true,
+	"$anchor": true,
+	"$recursiveAnchor": true,
+	"$defs": true,
+	"definitions": true,
+	"$ref": true,
+	"$recursiveRef": true,
+	// "title": true,
+	// "description": true,
+	// "default": true,
+	// "examples": true,
+	// "readOnly": true,
+	// "writeOnly": true,
+	// "$comment": true,
+	"$vocabulary": true,
+
+	// array keywords
+	// "items": true,
+	// "additionalItems": true,
+	// "maxItems": true,
+	// "minItems": true,
+	// "uniqueItems": true,
+	// "contains": true,
+	// "maxContains": true,
+	// "minContains": true,
+	"unevaluatedItems": true,
+
+	// object keywords
+	"unevaluatedProperties": true,
+
+	// conditional keywords
+	"if": true,
+	"then": true,
+	"else": true,
+
+	// other
+	"contentEncoding": true,
+	"contentMediaType": true,
+	"contentSchema": true,
+	"dependentRequired": true,
+	"dependentSchemas": true,
+	"deprecated": true,
+}
+
+var KeywordRegistry = map[string]KeyMaker{}
+var KeywordOrder = map[string]int{}
+
+func IsKeyword(prop string) bool {
+	_, ok := KeywordRegistry[prop]
+	return ok
+}
+
+func GetKeyword(prop string) Keyword {
+	if !IsKeyword(prop) {
+		return NewVoid()
+	}
+	return KeywordRegistry[prop]()
+}
+
+func GetKeywordOrder(prop string) int {
+	if order, ok := KeywordOrder[prop]; ok {
+		return order
+	}
+	return 1
+}
+
+func SetKeywordOrder(prop string, order int) {
+	KeywordOrder[prop] = order
+}
+
+func IsNotSupportedKeyword(prop string) bool {
+	_, ok := notSupported[prop]
+	return ok
+}
+
+func RegisterKeyword(prop string, maker KeyMaker) {
+	KeywordRegistry[prop] = maker
+}
+
+// MaxValueErrStringLen sets how long a value can be before it's length is truncated
+// when printing error strings
+// a special value of -1 disables output trimming
+var MaxKeywordErrStringLen = 20
+
+// Validator is an interface for anything that can validate.
+// JSON-Schema keywords are all examples of validators
+type Keyword interface {
+	// Validate checks decoded JSON data and writes
+	// validation errors (if any) to an outparam slice of ValErrors
+	// propPath indicates the position of data in the json tree
+	Validate(propPath string, data interface{}, errs *[]KeyError)
+	ValidateFromContext(schCtx *SchemaContext, errs *[]KeyError)
```

Comment:
Speaking of _context_, if possible avoid using the term. But if it makes the most sense, let's run with it